multisubscribe.go

```diff
@@ -0,0 +1,124 @@
+package legs
+
+import (
+	"context"
+	"os"
+	"sync/atomic"
+
+	dt "github.com/filecoin-project/go-data-transfer"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-graphsync"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/libp2p/go-libp2p-core/host"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	errors "golang.org/x/xerrors"
+)
+
+const (
+	// directConnectTicks makes pubsub check it's connected to direct peers every N seconds.
+	directConnectTicks uint64 = 30
+)
+
+// LegMultiSubscriber allows you to setup multiple subscriptions over single graphsync/datatransfer instance
+// with difference policy filters
+type LegMultiSubscriber interface {
+	NewSubscriber(PolicyHandler) (LegSubscriber, error)
+	NewSubscriberPartiallySynced(PolicyHandler, cid.Cid) (LegSubscriber, error)
+	GraphSync() graphsync.GraphExchange
+	DataTransfer() dt.Manager
+	Close(ctx context.Context) error
+}
+
+type legMultiSubscriber struct {
+	ctx    context.Context
+	tmpDir string
+	t      dt.Manager
+	gs     graphsync.GraphExchange
+	topic  *pubsub.Topic
+	refc   int32
+}
+
+// NewMultiSubscriber sets up a new instance of a multi subscriber
+func NewMultiSubscriber(ctx context.Context, host host.Host, ds datastore.Batching, lsys ipld.LinkSystem, topic string) (LegMultiSubscriber, error) {
```

Comment:
this is for use in the indexer, presumably. is this an interface that storetheindex will be able to integrate? since that's the downstream that we should probably update with this interface change

Reply:
Yes, we'll need to propagate these changes to storetheindex. I think they may be quite straightfoward once we implement `Fetch` in MultiSubscriber.